ray-curator/ray_curator/stages/shuffler/rapidsmpf_shuffler.py

```diff
@@ -0,0 +1,249 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.
+"""
+This module implements a bulk-synchronous shuffle class using UCXX communication compatible with Ray Actors.
+"""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Literal
+
+import cudf
+import pylibcudf as plc
+import rmm.mr
+from rapidsmpf.buffer.buffer import MemoryType
+from rapidsmpf.buffer.resource import BufferResource, LimitAvailableMemory
+from rapidsmpf.shuffler import partition_and_pack, unpack_and_concat
+from rapidsmpf.statistics import Statistics
+from rapidsmpf.utils.cudf import cudf_to_pylibcudf_table, pylibcudf_to_cudf_dataframe
+from rapidsmpf.utils.ray_utils import BaseShufflingActor
+
+from ray_curator.stages.deduplication.gpu_utils import align_down_to_256
+
+if TYPE_CHECKING:
+    from collections.abc import Iterator
+
+    from rapidsmpf.shuffler import Shuffler
+
+
+class BulkRapidsMPFShuffler(BaseShufflingActor):
+    """
+    Class that performs a bulk shuffle operation.
+    This class is compatible with Ray Actors communicating with each other using UCXX communication.
+    Parameters
+    ----------
+    nranks
+        Number of ranks in the communication group.
+    total_nparts
+        Total number of output partitions.
+    shuffle_on
+        List of column names to shuffle on.
+    output_path
+        Path to write output files.
+    rmm_pool_size
+        Size of the RMM memory pool in bytes.
+    spill_memory_limit
+        Device memory limit in bytes for spilling to host.
+        If "auto", the limit is set to 80% of the RMM pool size.
+        If None spilling is disabled.
+    enable_statistics
+        Whether to collect shuffle statistics.
+    """
+
+    def __init__(  # noqa: PLR0913
+        self,
+        nranks: int,
+        total_nparts: int,
+        shuffle_on: list[str],
+        output_path: str = "./",
+        rmm_pool_size: int = 1024 * 1024 * 1024,
+        spill_memory_limit: int | Literal["auto"] | None = "auto",
+        *,
+        enable_statistics: bool = False,
+    ):
+        super().__init__(nranks)
+        self.shuffle_on = shuffle_on
+        self.output_path = output_path
+        self.total_nparts = total_nparts
+        self.rmm_pool_size = align_down_to_256(rmm_pool_size)
+
+        if isinstance(spill_memory_limit, int):
+            self.spill_memory_limit = align_down_to_256(spill_memory_limit)
+        elif spill_memory_limit == "auto":
+            self.spill_memory_limit = align_down_to_256(0.8 * self.rmm_pool_size)
+        elif spill_memory_limit is None:
+            self.spill_memory_limit = None
+        else:
+            err_msg = f"Invalid spill_memory_limit: {spill_memory_limit}"
+            raise ValueError(err_msg)
```

Comment:
I realised I'm not using any of these things in Semantic, maybe I should.. will create a ticket